fix(common): date pipe is giving error weeks

ajitsinghkaler opened the pull request:
DDtae pipe is giving wrong week number made a change that will return correct week numbers for ifferent years

Fixes #33961

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -71,6 +71,12 @@ import {JitReflector} from '@angular/platform-browser-dynamic/src/compiler_refle
       it('should use "mediumDate" as the default format',
          () => expect(pipe.transform('2017-01-11T10:14:39+0000')).toEqual('Jan 11, 2017'));
+
+      it('should return first week if some dates fall in previous year but belong to next year accordint to ISO 8601 format', () => {
```
Drive-by suggestion:

```suggestion
      it('should return first week if some dates fall in previous year but belong to next year according to ISO 8601 format', () => {
```

ajitsinghkaler replied:
@petebacondarwin thanks made the change

ajitsinghkaler committed:
fix(common): date pipe gives wrong week number

Date pipe is giving wrong week number when used with the date format 'w'. If first week(according to Iso)  has some days in previous year

Fixes #33961

ajitsinghkaler commented:
@AndrewKushnir flaky failures on ci

AndrewKushnir commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -382,8 +382,11 @@ function weekGetter(size: number, monthBased = false): DateFormatter {
-      const firstThurs = getFirstThursdayOfYear(date.getFullYear());
+      let firstThurs = getFirstThursdayOfYear(date.getFullYear());
       const thisThurs = getThursdayThisWeek(date);
+      if (thisThurs.getFullYear() !== firstThurs.getFullYear()) {
```
Could you please add a comment here describing this logic?

AndrewKushnir commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -71,6 +71,11 @@ import {JitReflector} from '@angular/platform-browser-dynamic/src/compiler_refle
+
+      it('should return first week if some dates fall in previous year but belong to next year according to ISO 8601 format',
+         () => {
+           expect(pipe.transform('2019-12-29T00:00:00', 'w')).toEqual('1');
```
I'd propose to extend this test to include a couple more days (2019-12-28 and 2019-12-30), so that it's clear when the switch from `52` to `1` actually happens. Also may be add the same checks for a leap year (2016).

AndrewKushnir reviewed (commented):
Hi @ajitsinghkaler, thanks for the fix 👍 

I've left a couple comments, please have a look when you get a chance.

FYI I've added the "breaking change" label to this PR. Even though it's a bug fix, it changes the behavior that other developers may rely on.

I'd also like to ask @petebacondarwin to have a look at this change as I'm not super familiar with this code and having additional reviewer would be very helpful.

Thank you.

ajitsinghkaler committed:
fixup! fix(common): date pipe gives wrong week number

petebacondarwin commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -382,8 +382,16 @@ function weekGetter(size: number, monthBased = false): DateFormatter {
       result = 1 + Math.floor((today + nbDaysBefore1stDayOfMonth) / 7);
     } else {
-      const firstThurs = getFirstThursdayOfYear(date.getFullYear());
+      let firstThurs = getFirstThursdayOfYear(date.getFullYear());
       const thisThurs = getThursdayThisWeek(date);
+      /*
+       If some days of a year are part of next year according
+       to ISO 8601. Change the firstThurs for those days to next
+       year firstThurs
+      */
+      if (thisThurs.getFullYear() !== firstThurs.getFullYear()) {
+        firstThurs = getFirstThursdayOfYear(date.getFullYear() + 1);
+      }
```
I believe we can make this logic much simpler...

```suggestion
      const thisThurs = getThursdayThisWeek(date);
      // Some days of a year are part of next year according to ISO 8601.
      // Compute the firstThurs from the year of this week's Thursday
      const firstThurs = getFirstThursdayOfYear(thisThurs.getFullYear());
```

WDYT? Am I missing something?

ajitsinghkaler replied:
seems much better to me

petebacondarwin reviewed (changes requested):
Thanks for this fix. It appears to be correct. 
I suggest a refactoring that would be less code and a bit simpler.

AndrewKushnir commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -71,6 +71,20 @@ import {JitReflector} from '@angular/platform-browser-dynamic/src/compiler_refle
+      it('should return first week if some dates fall in previous leap year but belong to next year according to ISO 8601 format',
+         () => {
+           expect(pipe.transform('2012-12-29T00:00:00', 'w')).toEqual('52');
+           expect(pipe.transform('2019-12-30T00:00:00', 'w')).toEqual('1');
```
@ajitsinghkaler thanks for adding more test cases!

It looks like the year is wrong in this case and in the one below (2019 instead of 2012).

ajitsinghkaler committed:
fixup! fix(common): date pipe gives wrong week number

petebacondarwin reviewed (approved):
LGTM

AndrewKushnir commented:
FYI I've started [presubmit](http://test/OCL:317675331:BASE:317675367:1592845083327:b67ffbd0) in Google's codebase, will let you know the result as soon as I have it. Thank you.

AndrewKushnir reviewed (approved):
Thanks for the fix @ajitsinghkaler 👍

AndrewKushnir commented:
Hi @ajitsinghkaler, just want to give a quick update. Running tests in Google's codebase didn't indicate any issues, which is great. Also given the fact that this is a bug fix with low risk, I'm removing the "breaking changes" flag. In any case, since the behavior of the framework changes slightly (to become more accurate), we would merge it to master only and it will be released as a part of the next minor version (v10.1.0). Thank you.
